curriculum/challenges/english/03-front-end-libraries/react/use-state-to-toggle-an-element.english.md

```diff
@@ -30,6 +47,8 @@ tests:
     testString: assert.strictEqual(Enzyme.mount(React.createElement(MyComponent)).state('visibility'), false);
   - text: Clicking the button element should toggle the <code>visibility</code> property in state between <code>true</code> and <code>false</code>.
     testString: 'async () => { const waitForIt = (fn) => new Promise((resolve, reject) => setTimeout(() => resolve(fn()), 250)); const mockedComponent = Enzyme.mount(React.createElement(MyComponent)); const first = () => { mockedComponent.setState({ visibility: false }); return waitForIt(() => mockedComponent.state(''visibility'')); }; const second = () => { mockedComponent.find(''button'').simulate(''click''); return waitForIt(() => mockedComponent.state(''visibility'')); }; const third = () => { mockedComponent.find(''button'').simulate(''click''); return waitForIt(() => mockedComponent.state(''visibility'')); }; const firstValue = await first(); const secondValue = await second(); const thirdValue = await third(); assert(!firstValue && secondValue && !thirdValue); }; '
+  - text: The <code>setState</code> version used should take a function.
+    testString: assert(new RegExp('this\\.setState\\(.*(=>|function)').test(code));
```

Comment:
I'm currently thinking of changing this line, as it would make the tests fail if the user defines the function first and then passes it to `setState` (or if the user formats the code weirdly).
```suggestion
    testString: assert(code.includes("=>") || code.includes("function"));
```
Alternatively, I can make the test description more descriptive:
> `function` or `=>` should be on the same line as `setState`

Reply:
How about "Pass an anonymous function to <code>setState</code> to..." in the instructions?

Reply:
@ojeytonwilliams I think that's a good idea but there's still the formatting issue:
```js
// wouldn't work
this.setState(
  s => ({visibility: !state.visibility})
);
```
However, I've written a better regex:
```js
/this\.setState\s*\(\s*(function|\(?\s*[a-zA-Z\s,_]+\s*\)?\s*=>)/
```

Reply:
Nice one!  Could you commit those changes?  Then I'll verify it works as intended.

Reply:
@ojeytonwilliams I've commited these changes.

Reply:
Thanks, checking it now.  By the way, if you want to ignore spaces, it can be easier simply to remove them from the code with `replace`, rather than allow for each one in the regex.  That's not a requirement at all, just a shortcut you can use if you want.